sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -79,7 +89,34 @@ class GpuHashAggregateMeta(
         case "partial" => if (hashAggMode.contains(Final) || hashAggMode.contains(Complete)) {
           // replacing only Partial hash aggregates, so a Final or Complete one should not replace
           willNotWorkOnGpu("Replacing Final or Complete hash aggregates disabled")
-         }
+        }
+          // In partial mode, if there are non-distinct functions and multiple distinct functions,
+          // non-distinct functions are computed using the First operator. The final result would be
+          // incorrect for non-distinct functions for partition size > 1. Reason for this is - if
+          // the first batch computed and sent to CPU doesn't contain all the rows required to
+          // compute non-distinct function(s), then Spark would consider that value as final result
+          // (due to First). Fall back to CPU in this case.
+          if (agg.aggregateExpressions.exists(e => e.aggregateFunction.isInstanceOf[First])) {
+            agg.aggregateExpressions.foreach(e => {
+              // Check if there is an `If` within `First`. This is included in the plan only when
+              // multiple distinct is present in the query. We fall back to CPU in this case when
+              // the references of `If` is an aggregate. We cannot call `isDistinct` here on
+              // aggregateExpressions to get the total number of distinct functions. If there are
+              // multiple distinct the plan is rewritten by `RewriteDistinctAggregates` where
+              // regular aggregations and every distinct aggregation is calculated in a separate
+              // group.
+              e.aggregateFunction match {
```

Comment:
I think all of this can just be rewritten to.

```
e.aggregateFunction match {
  case First(If, _) =>
      ...
   case _ =>
}
```

And I think we can combine all of these together in the exists and the foreach, but it would take a min or two to do it.